src/Theme.cpp

```diff
@@ -0,0 +1,411 @@
+#include "stdafx.h"
+
+#include <fmt/core.h>
+#include <scn/scn.h>
+
+#include "Utils.h"
+
+#include "Theme.h"
+
+using nlohmann::json;
+
+// Helpers
+
+static json XyDimensionsToObject(const ImVec2& aXy)
+{
+    return json::object({{"x", aXy.x}, {"y", aXy.y}});
+}
+
+static ImVec2 XyDimensionsFromObject(const json& aJson, const ImVec2& aDefaults)
+{
+    return ImVec2(aJson.value("x", aDefaults.x), aJson.value("y", aDefaults.y));
+}
+
+static json DirectionToLabel(const ImGuiDir& aDirection)
+{
+    switch (aDirection)
+    {
+    case ImGuiDir_None: return "None";
+    case ImGuiDir_Left: return "Left";
+    case ImGuiDir_Right: return "Right";
+    case ImGuiDir_Up: return "Up";
+    case ImGuiDir_Down: return "Down";
+    default: return "Unknown";
+    }
+}
+
+static ImGuiDir DirectionFromLabel(const std::string& aDirectionLabel, const ImGuiDir& aDefaultDirection)
+{
+    if (aDirectionLabel == "None")
+        return ImGuiDir_None;
+    if (aDirectionLabel == "Left")
+        return ImGuiDir_Left;
+    if (aDirectionLabel == "Right")
+        return ImGuiDir_Right;
+    if (aDirectionLabel == "Up")
+        return ImGuiDir_Up;
+    if (aDirectionLabel == "Down")
+        return ImGuiDir_Down;
+
+    return aDefaultDirection;
+}
+
+static std::string ColorToHex(const ImVec4& aColor)
+{
+    return fmt::format(
+        //    R     G     B     A
+        "#{:02x}{:02x}{:02x}{:02x}", static_cast<int>(aColor.x * 255), static_cast<int>(aColor.y * 255), static_cast<int>(aColor.z * 255), static_cast<int>(aColor.w * 255));
+}
+
+static ImVec4 ColorFromHex(const std::string& aHex, const ImVec4& aDefaultColor)
+{
+    if (aHex.empty())
+    {
+        return aDefaultColor;
+    }
+
+    int r, g, b = 0;
+
+    auto foundRGB = scn::scan(aHex, "#{:2x}{:2x}{:2x}", r, g, b);
+
+    if (!foundRGB)
+    {
+        Log::Warn("Failed to parse color, must be hex with optional alpha (\"#00ff00\", \"#00ff00ff\"): {}", aHex);
+        return aDefaultColor;
+    }
+
+    int a = 255;
+    auto alphaResult = scn::scan(foundRGB.range(), "{:2x}", a);
+
+    ImVec4 outColor{r / 255.0f, g / 255.0f, b / 255.0f, a / 255.0f};
+
+    return outColor;
+}
+
+//
+// Codec
+//
+
+// Encoders
+using EncodeValueFunc = std::function<void(const ImGuiStyle&, json& aOutJson)>;
+
+// Decoders
+using DecodeValueFunc = std::function<void(ImGuiStyle& aOutStyle, const json& aJson)>;
+
+struct Codec
+{
+    std::string valueName;
+    EncodeValueFunc encodeFrom;
+    DecodeValueFunc decodeInto;
+};
+
+//
+// Simple scalar types
+//
+
+template <typename TAccessor> static EncodeValueFunc EncodeScalar(std::string aValueName, TAccessor aValueAccessor)
```

Comment:
const std::string& aValueName
dont unnecessarily copy
= in lambda is probably fine as you want to make a copy there, but you do 2 instead of 1 this way